include/SFML/System/UniqueResource.hpp

```diff
@@ -0,0 +1,160 @@
+template <typename T, typename Deleter>
+class UniqueResource
+{
+public:
+    ////////////////////////////////////////////////////////////
+    /// \brief Default constructor
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource();
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Construct from resource and deleter
+    ///
+    /// \param resource Resource to store
+    /// \param deleter Deleter for the resource
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource(T&& resource, Deleter&& deleter);
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Delete the resource
+    ///
+    ////////////////////////////////////////////////////////////
+    ~UniqueResource();
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Deleted copy constructor
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource(const UniqueResource&) = delete;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Deleted copy assignment
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource& operator=(const UniqueResource&) = delete;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Move constructor
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource(UniqueResource&& other) noexcept;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Move assignment
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource& operator=(UniqueResource&& other) noexcept;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Release ownership of the managed resource
+    ///
+    ////////////////////////////////////////////////////////////
+    void release();
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Dispose of the resource by calling the deleter
+    ///
+    ////////////////////////////////////////////////////////////
+    void reset();
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Dispose of the resource by calling the deleter
+    /// then assign to the resource
+    ///
+    ////////////////////////////////////////////////////////////
+    void reset(T&& resource);
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Access the underlying resource
+    ///
+    ////////////////////////////////////////////////////////////
+    [[nodiscard]] const T& get() const;
+
+private:
+    T       m_resource; //!< Handle to the managed resource
+    Deleter m_deleter;  //!< Functor to clean up resource
+    bool    m_delete;   //!< Whether or not a resource is currently owned
```

Comment:
`m_delete` is not really a clear name, especially since we already have `m_deleter` -- better would be something like `m_isOccupied`.

Reply:
I swapped this out for an `optional<Deleter>` and I think the implementation looks better now so I just sidestepped that naming question.